aptos-move/block-executor/src/executor.rs

```diff
@@ -655,32 +712,36 @@ where
         txn_idx: TxnIndex,
         versioned_cache: &MVHashMap<T::Key, T::Tag, T::Value, DelayedFieldID>,
         last_input_output: &TxnLastInputOutput<T, E::Output, E::Error>,
+        is_v2: bool,
     ) -> Result<bool, PanicError> {
         let read_set = last_input_output
             .read_set(txn_idx)
             .expect("Read set must be recorded");
 
-        let mut execution_still_valid =
-            read_set.validate_delayed_field_reads(versioned_cache.delayed_fields(), txn_idx)?;
+        if !read_set.validate_delayed_field_reads(versioned_cache.delayed_fields(), txn_idx)?
+            || (is_v2
+                && !read_set.validate_aggregator_v1_reads(
+                    versioned_cache.data(),
+                    last_input_output.modified_aggregator_v1_keys(txn_idx),
+                    txn_idx,
+                )?)
+        {
+            return Ok(false);
+        }
 
-        if execution_still_valid {
-            if let Some(delayed_field_ids) = last_input_output.delayed_field_keys(txn_idx) {
-                if let Err(e) = versioned_cache
-                    .delayed_fields()
-                    .try_commit(txn_idx, delayed_field_ids.collect())
-                {
-                    match e {
-                        CommitError::ReExecutionNeeded(_) => {
-                            execution_still_valid = false;
-                        },
-                        CommitError::CodeInvariantError(msg) => {
-                            return Err(code_invariant_error(msg));
-                        },
-                    }
-                }
+        if let Some(delayed_field_ids) = last_input_output.delayed_field_keys(txn_idx) {
+            if let Err(e) = versioned_cache
+                .delayed_fields()
+                .try_commit(txn_idx, delayed_field_ids.collect())
```

Comment:
nit: why collect? We already have an iter, we can probably just pass it here avoiding new vec creation?